### PR TITLE
Add a fix to Groupby for aggregations by a column from the DataFrame

### LINKED ISSUE #409: Getting an error when I try to call an aggregate on a groupby object
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **Modin installed from (source or binary)**: binary
- **Modin version**: 0.2.5
- **Python version**: 3.6.8
- **Exact command to reproduce**: 
```python
import modin.pandas as pd
df = pd.read_csv(file)
df.groupby(col).size()
```

### Describe the problem
Getting an error when I try to call .size or other aggregates like .count on a groupby object

### Source code / logs

```
~/miniconda3/envs/elo-kaggle/lib/python3.6/site-packages/modin/pandas/dataframe.py in __repr__(self)
    122         num_cols = 30
    123 
--> 124         result = repr(self._build_repr_df(num_rows, num_cols))
    125         if len(self.index) > num_rows or len(self.columns) > num_cols:
    126             # The split here is so that we don't repr pandas row lengths.

~/miniconda3/envs/elo-kaggle/lib/python3.6/site-packages/modin/pandas/dataframe.py in _build_repr_df(self, num_rows, num_cols)
     93 
     94         if len(self.columns) <= num_cols:
---> 95             head_front = head.to_pandas()
     96             # Creating these empty to make the concat logic simpler
     97             head_back = pandas.DataFrame()

~/miniconda3/envs/elo-kaggle/lib/python3.6/site-packages/modin/data_management/query_compiler/pandas_query_compiler.py in to_pandas(self)
   2047             df = pandas.DataFrame(dtype_dict, self.index)
   2048         else:
-> 2049             df.index = self.index
   2050             df.columns = self.columns
   2051         return df

~/miniconda3/envs/elo-kaggle/lib/python3.6/site-packages/pandas/core/generic.py in __setattr__(self, name, value)
   4387         try:
   4388             object.__getattribute__(self, name)
-> 4389             return object.__setattr__(self, name, value)
   4390         except AttributeError:
   4391             pass

pandas/_libs/properties.pyx in pandas._libs.properties.AxisProperty.__set__()

~/miniconda3/envs/elo-kaggle/lib/python3.6/site-packages/pandas/core/series.py in _set_axis(self, axis, labels, fastpath)
    387         object.__setattr__(self, '_index', labels)
    388         if not fastpath:
--> 389             self._data.set_axis(axis, labels)
    390 
    391     def _set_subtyp(self, is_all_dates):

~/miniconda3/envs/elo-kaggle/lib/python3.6/site-packages/pandas/core/internals.py in set_axis(self, axis, new_labels)
   3321             raise ValueError(
   3322                 'Length mismatch: Expected axis has {old} elements, new '
-> 3323                 'values have {new} elements'.format(old=old_len, new=new_len))
   3324 
   3325         self.axes[axis] = new_labels

ValueError: Length mismatch: Expected axis has 186 elements, new values have 31 elements
```

### PR DESCRIPTION
**Pull request by devin-petersohn**
* Resolves #409
* Removes a grouped column from the result to match pandas
* Changes the way we compute `size` to match pandas
* Adds consisntecy between the DataFrame being computed on and the result

## Related issue number

- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] passes `black --check modin/`

### CODE REVIEW

**Comment by devin-petersohn** [comment]
Hey @eavidan, that sounds good. I think dropping the non-numeric columns is what we try to do, but `sum` allows for string sums (concatenation). So we assume that `numeric` operations can only happen on numeric columns, but `sum` allows strings as well. Code [here](https://github.com/modin-project/modin/blob/master/modin/pandas/groupby.py#L439-L444).

Truthfully I need to spend some time on the tests for this code path because we aren't testing enough of the edge cases. I should add that to this PR so we can make sure that this and all future updates to this code are covered well.

File: modin/data_management/query_compiler/pandas_query_compiler.py [hunk 1]
```diff
@@ -2368,12 +2368,12 @@ def _post_process_apply(self, result_data, axis, try_scale=True):
             else:
                 columns = self.columns
         else:
-            # See above explaination for checking the lengths of columns
+            columns = internal_columns
```
Comment (devin-petersohn) [inline modin/data_management/query_compiler/pandas_query_compiler.py]: I moved this for consistency with code block above.

File: modin/data_management/query_compiler/pandas_query_compiler.py [hunk 2]
```diff
@@ -2492,20 +2491,32 @@ def groupby_agg(self, by, axis, agg_func, groupby_args, agg_args):
...
+                df.index = pandas.RangeIndex(len(df.index))
             grouped_df = df.groupby(by=by, axis=axis, **groupby_args)
             try:
                 return agg_func(grouped_df, **agg_args)
             # This happens when the partition is filled with non-numeric data and a
             # numeric operation is done. We need to build the index here to avoid issues
             # with extracting the index.
             except DataError:
-                return pandas.DataFrame(index=grouped_df.count().index)
+                return pandas.DataFrame(index=grouped_df.size().index)
```
Comment (devin-petersohn) [inline modin/data_management/query_compiler/pandas_query_compiler.py]: Changed to `size` which has lower overhead that `count`.

File: modin/pandas/test/test_groupby.py [hunk 3]
```diff
@@ -215,7 +296,6 @@ def test_single_group_row_groupby():
     test_take(ray_groupby, pandas_groupby)
 
 
-@pytest.mark.skip(reason="See Modin issue #21.")
```
Comment (devin-petersohn) [inline modin/pandas/test/test_groupby.py]: Was able to add this test back. It was working before this PR also.

**Comment by devin-petersohn** [comment]
@eavidan, I fixed the issue for the general case and added some test cases for larger and variable dtype DataFrames.

File: modin/data_management/query_compiler/pandas_query_compiler.py [hunk 4]
```diff
@@ -2492,20 +2491,32 @@ def groupby_agg(self, by, axis, agg_func, groupby_args, agg_args):
...
 
         func_prepared = self._prepare_method(lambda df: groupby_agg_builder(df))
         result_data = self.map_across_full_axis(axis, func_prepared)
-        return self._post_process_apply(result_data, axis, try_scale=False)
+        if axis == 0:
```
[the remark this answers is not in the document]
Comment (devin-petersohn) [reply modin/data_management/query_compiler/pandas_query_compiler.py]: We don't have any way of knowing `if len(columns) == 0` before this. This is how we check.

[the remark this answers is not in the document]
Comment (williamma12) [reply modin/data_management/query_compiler/pandas_query_compiler.py]: Ah sorry, my mistake

**Comment by eavidan** [comment]
@devin-petersohn great PR and nice variety of tests. 
Pandas seems to have an `numeric_only` passed as one of `sum`s `kwargs`, where you can state if you want to sum only numeric columns or not. 
It seems that those `kwargs` we pass to the lamda function in `_apply_agg_function()` do not reach `pandas.sum()` at the partition level when we `map_across_full_axis`.
So we still get some inconsistencies in some cases. eg:
```py
df = pd.DataFrame({'A': [2, 2, 2, 3, 4],
                    'B': [5, 6, 7, 8, 9],
                    'C': ['a', 'b', 'c', 'd', 'e']})

df = df.groupby('A').sum(numeric_only=True)
print(df)
```
This returns with the `C` column where pandas do not (`numeric_only` is set to `True` by default).

**Comment by devin-petersohn** [comment]
Thanks @eavidan, I see what is happening. Pandas will ignore `numeric_only` if there are no numeric columns in the DataFrame.

```python
import pandas
df = pandas.DataFrame({'A': [2, 2, 2, 3, 4],
                    'B': [5, 6, 7, 8, 9],
                    'C': ['a', 'b', 'c', 'd', 'e']})

df = df[["A", "C"]].groupby('A').sum(numeric_only=True)
print(df)
```
It is strange behavior. For now, since this fixes something very broken, we can merge this. I will create a new PR to handle the `kwargs` differently.
